docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+Date:   Mon Oct 10 13:25:51 2016 -0700
+
+    Merge branch 'master' of https://github.com/zulip/zulip
+```
+
+To undo a merge commit, first run `git reflog` to identify the commit you wanto to roll back to:
+
+```
+$ git reflog
+
+e5f8211 HEAD@{0}: pull upstream master: Merge made by the 'recursive' strategy.
+13bea0e HEAD@{1}: commit: test commit for docs.
+```
+
+Reflog output will be long. The most recent git refs will be listed at the top. In the example above `e5f8211 HEAD@{0}:` is the merge commit made automatically by `git pull` and `13bea0e HEAD@{1}:` is the last commit I made before running `git pull`, the commit that I want to rollback to.
+
+Once you'd identified the ref you want to revert to, you can do so with [git reset][gitbook-reset]:
+
+```
+$ git reset --hard 13bea0e
+HEAD is now at 13bea0e test commit for docs.
+```
+
+**Important:** `git reset --hard <commit>` will discard all changes in your working directory and index since the commit you're resetting to with *<commit>*. If you need to keep any changes that are in your working directory or that you have commmited, use `git reset --merge <commit>` instead.
+
+You can also use the relative refleg `HEAD@{1}` intead of the commit hash, just keep in mind this changes as you run git commands.
+
+Now when I look at the git reflog, I see the tip of my branch is pointing to my last commit `13bea0e` before the merge:
+
+```
+$ git reflog
+
+13bea0e HEAD@{2}: reset: moving to HEAD@{1}
+e5f8211 HEAD@{3}: pull upstream master: Merge made by the 'recursive' strategy.
+13bea0e HEAD@{4}: commit: test commit for docs.
+```
+
+And the first entry `git log` shows is this:
+
+```
+commit 13bea0e40197b1670e927a9eb05aaf50df9e8277
+Author: Christie Koehler <ck@christi3k.net>
+Date:   Mon Oct 10 13:25:38 2016 -0700
+
+    test commit for docs.
+```
+
+### Restore a lost commit
+
+We've mentioned you can use `git reset --hard` to rollback to a previous commit. What if you run `git reset --hard` and then realize you actually need one or more of the commits you just disgarded? No problem, you can restore them with `git merge`.
+
+For example, let's say you just commited "some work" and your `git log` looks like this:
+
+```
+* 67aea58 (HEAD -> master) some work
+* 13bea0e test commit for docs.
+```
+
+You then mistakenly run `git reset --hard 13bea0e`:
+
+```
+$ git reset --hard 13bea0e
+HEAD is now at 13bea0e test commit for docs.
+
+$ git log
+* 13bea0e (HEAD -> master) test commit for docs.
+```
+
+And then realize you actually needed to keep commit 67aea58. First, use `git reflog` to confirm that commit you want to restore and then run `git merge <commit>`:
+
+```
+$ git reflog
+13bea0e HEAD@{0}: reset: moving to 13bea0e
+67aea58 HEAD@{1}: commit: some work
+
+$ git merge 67aea58
+Updating 13bea0e..67aea58
+Fast-forward
+ test3.txt | 1 +
+ test4.txt | 1 +
+ 2 files changed, 2 insertions(+)
+ create mode 100644 test4.txt
+```
+
+### Recover from a git rebase failure
```

Comment:
Oh, this reminds me -- we should recommend that people setup their `~/.gitconfig` properly, e.g. here's mine:

```
 cat ~/.gitconfig 
[user]
	name = Tim Abbott
	email = tabbott@zulipchat.com
[core]
	editor = emacs -nw
[merge]
	conflictstyle = diff3
```

The conflictstyle thing causes it to produce a nicer 3-way diff.